crowbar_framework/lib/crowbar/chef/upload.rb

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env ruby
+#
+# Copyright 2011-2013, Dell
+# Copyright 2013-2016, SUSE Linux GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/cookbook_uploader"
+require "chef/knife/core/object_loader"
+
+module Crowbar
+  class Chef
+    class Upload
+      class << self
+        def all
+          cookbooks
+          data_bags
+          roles
+        end
+
+        def cookbooks
+          version_constraints_to_update = Hash.new
+          cookbooks_for_upload = Array.new
+
+          cookbooks_repo = ::Chef::CookbookLoader.new(chef_cookbooks_path)
+          cookbooks_repo.each do |cookbook_name, cookbook|
+            cookbooks_for_upload.push(cookbook)
+            version_constraints_to_update[cookbook_name] = cookbook.version
+          end
+
+          if cookbooks_for_upload.any?
+            cookbooks_for_upload.each do |cookbook|
+              logger.info("Checking cookbook #{cookbook.name}")
+              return false unless check_links(cookbook)
+              return false unless check_dependencies(cookbook)
+            end
+
+            cookbook_names = cookbooks_for_upload.map do |cookbook|
+              "#{cookbook.name}"
+            end
+
+            logger.info("Uploading cookbooks... #{cookbook_names.join(", ")}")
+            ::Chef::CookbookUploader.new(
+              cookbooks_for_upload,
+              force: false,
+              concurrency: true
+            ).upload_cookbooks
+          end
+        end
+
+        def data_bags
```

Comment:
To be honest, I think we should stop uploading the data bags, they shouldn't be needed as data bags anymore. I'd work on fixing that instead.